[a11y] Fix issue with iOS not reading updated header

bkjohnson opened the pull request:
## Description

~~This removes the environment flags from #14024 , so the fix for https://github.com/department-of-veterans-affairs/va.gov-team/issues/12323 will be available in production.~~

~~Note: this means that the styling change in #14036 will also take effect in production.~~

This PR has been repurposed to get the changes from #14183 into staging for further testing. These changes are aimed to remove the header and re-render it on a chapter change so that VoiceOver on iOS will notice the change.


## Testing done

@1Copenut has done testing in staging to validate that this works as expected.


## Screenshots

Here is an example of the re-render happening - I've changed the display text to make it a bit more obvious, and here is the diff:

```diff
diff --git a/src/platform/forms-system/src/js/components/FormNav.jsx b/src/platform/forms-system/src/js/components/FormNav.jsx
index 863844b259..68683ba859 100644
--- a/src/platform/forms-system/src/js/components/FormNav.jsx
+++ b/src/platform/forms-system/src/js/components/FormNav.jsx
@@ -66,6 +66,7 @@ export default function FormNav(props) {
   // https://github.com/department-of-veterans-affairs/va.gov-team/issues/12323
   useEffect(
     () => {
+      debugger;
       if (!environment.isProduction()) {
         if (current > index + 1) {
           setIndex(index + 1);
@@ -101,9 +102,7 @@ export default function FormNav(props) {
               </h2>
             )}
           {!environment.isProduction() &&
-            !showHeader && (
-              <div className="vads-u-font-size--h4">{stepText}</div>
-            )}
+            !showHeader && <div className="vads-u-font-size--h4">Testing</div>}
 
           {environment.isProduction() && (
             <h2 className="vads-u-font-size--h4">
```

![Peek 2020-09-15 11-27](https://user-images.githubusercontent.com/2008881/93249749-a36fb580-f746-11ea-9e3f-1ca2ed92f022.gif)




## Acceptance criteria
- [x] VoiceOver on iOS picks up the changed text in the new `h2`

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
Remove environment flag

1Copenut reviewed (approved):
LGTM. @pjhill and I are testing various forms in Staging manually. Not seeing an regressions, and have confirmed this fix for major screen reader + browser pairings.

bkjohnson commented:
We are planning on releasing this on Monday, 8/31

1Copenut commented:
@bkjohnson Can we hold on this one more day? I'd like to find a way to fix this iOS (new) bug that's causing it to read the first page heading on every page, instead of the correct heading.

/cc @meganhkelley && @pjhill

bkjohnson commented:
Sure, we can wait.

bkjohnson committed:
Use const for text display value

bkjohnson committed:
Merge branch 'master' into 12323-env-flag-removal

bkjohnson committed:
[a11y] Fix iOS VoiceOver bug (#14183)

* Convert FormNav to a functional component

This uses React.memo to achieve what the shouldComponentUpdate lifecycle
method was doing:

https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-shouldcomponentupdate

* Fix some failing tests by switching to RTL

* Remove performance optimization

The React.memo wasjust implementing what shouldComponentUpdate was doing
previously, but this is causing problems with some unit tests. Removing
it for now - maybe we'll add it back in later

* Try to sneakily remove & re-render the header

* Try to remove & re-render the h2

This hook-based forceUpdate is based on this:

https://reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate

I see the h2 being removed correctly, but it stays removed instead of
being immediately re-rendered

* Try to fix forceUpdate setup

* Use state changes to manipulate the h2 rendering

* Replace "Loading..." with stepText

* Focus the form header after the sneaky re-render

This is what the FormPage component does on each mount

* Use a more appropriate/meaningful name

bkjohnson commented:
@1Copenut, can you test this branch and see if it works as expected regarding the iOS bug?

cvalarida commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -11,82 +9,96 @@ import {
-          ? formConfig?.customText?.reviewPageTitle ||
-            REVIEW_APP_DEFAULT_MESSAGE
-          : formConfig.chapters[page.chapterKey].title;
-      if (typeof chapterName === 'function') {
-        chapterName = chapterName();
+  // The goal with this is to quickly "remove" the header from the DOM, and
+  // immediately re-render the component with the header included. This should
+  // ensure that VoiceOver on iOS will pick up on the new <h2>
+  useEffect(
+    () => {
+      if (current > index + 1) {
+        setIndex(index + 1);
+      } else if (current === index) {
+        setIndex(index - 1);
       }
```
This is a little confusing. Can you add a little clarity in the comments around how this works? I think calling out that `current` is changed when the chapter changes, and that when the chapter changes, we need to remove the heading element and re-display it so iOS will announce the new element's text. Bonus points for linking to the ticket.

cvalarida reviewed (approved):
I think my main concern with this now is that the `<div>` will be shown and won't be immediately removed to render the `<h2>`. I don't have time to test right now, but did you confirm that the `<div>` is only fleeting? (If we're missing a header for long enough for a user to notice, screen reader users may have trouble with header navigation.)

That said, I don't want to block you on this. If you confirm that the `<h2>` is only gone for an instant, it looks great! I just had one ask for a clarifying code comment.

1Copenut commented:
@bkjohnson I just tried this latest fix out, and it works with iOS and MacOS Safari + VO as expected! 🎉 

I'll need to test with the other screen reader pairings in Staging before we ship to production, but it *should* work as expected. LMK when we're in staging and I"ll fast-track the testing on JAWS and NVDA.

bkjohnson commented:
@cvalarida I was only able to spot the`<div>` when I had the debugger on - it happened so fast under normal execution that it wasn't obvious that something was being added and then "removed"

@1Copenut I'll add the environment flags back in here and we can send it to staging.

bkjohnson committed:
Revert "Remove environment flag"

This reverts commit 99bf8c45df4d0f0b0d56e7189daac342d7d04fce.

I made some further changes to FormNav to allow this

bkjohnson committed:
Add some comments

bkjohnson committed:
Only do the focus on cleaup in non-prod envs

The focusing in FormPage should still work fine since we aren't
re-rendering to remove the h2

bkjohnson committed:
Add another environment flag

I don't want these unnecessary state changes (and re-renderings)
happening in prod

bkjohnson commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -11,82 +10,112 @@ import {
+          className="nav-header nav-header-schemaform"
+        >
+          {!environment.isProduction() &&
+            showHeader && (
+              <h2 id="nav-form-header" className="vads-u-font-size--h4">
+                {stepText}
               </h2>
             )}
-          </div>
+          {!environment.isProduction() &&
+            !showHeader && (
+              <div className="vads-u-font-size--h4">{stepText}</div>
+            )}
+
+          {environment.isProduction() && (
+            <h2 className="vads-u-font-size--h4">
+              <span className="form-process-step current">{current}</span>{' '}
+              <span className="form-process-total">of {chapters.length}</span>{' '}
+              {chapterName}
+            </h2>
+          )}
```
This is admittedly awkward, but it should be very temporary. Lines 97-106 are effectively a ternary for non-prod environments. This "ternary" will show the actual `<h2>` under normal circumstances, but when we manipulate the `index` state variable to be far enough away from `current`, that means that we don't want to render the `h2` and will render the fake `div` header instead.

bkjohnson commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -11,82 +10,112 @@ import {
-            }: ${chapterName}`}
-            aria-valuemax={chapters.length}
-            className="nav-header nav-header-schemaform"
-          >
-            {!environment.isProduction() ? (
-              <h2 className="vads-u-font-size--h4">
-                {`Step ${current} of ${chapters.length}: ${chapterName}`}
-              </h2>
-            ) : (
-              <h2 className="vads-u-font-size--h4">
-                <span className="form-process-step current">{current}</span>{' '}
-                <span className="form-process-total">of {chapters.length}</span>{' '}
-                {chapterName}
+      return () => {
+        if (!environment.isProduction()) {
+          focusElement('.nav-header > h2');
+        }
+      };
+    },
+    [current, index],
```
These environment flags should prevent these effects from doing anything on production
